### ISSUE TITLE
Not compatible to XC16 compiled ELF files

### ISSUE THREAD

**Issue by yashagarwal-314** [eliben/pyelftools#473]

**Comment by yashagarwal-314** [comment]
I am trying to use the pyelftools to get the variable data type for the dspic33ck elf file and it doesn't work.

**Comment by sevaa** [comment]
Can you share one of the offending files? If they are sensitive, can you build a dummy project with the same toolchain and settings and, if the problem reproduces, share that?

Also, please share what are you doing exactly with pyelftools. Does the file not open, or something goes wrong later, during debug info parsing/navigation?

**Comment by yashagarwal-314** [comment]
Hello,

thank you for your email, the file opens but later I cannot access the
debug information, somehow I get an assertation error.
I am attaching the elf file of my project, hope that helps and thank you
for your help :)

I wish you a great day and looking forward to hearing from you soon.
Freundliche grüße
Yash Agarwal

**Comment by sevaa** [comment]
The attachment didn't go through. Please navigate to the issue at Github and attach it there.

**Comment by yashagarwal-314** [comment]
[https://drive.google.com/file/d/1oTaSIvRCXfMlEzVjOPPxdk4UA4pS4Bm_/view?usp=drive_link](url)

Hello @sevaa, 

Thank you for your message, I have uploaded it on G-Drive and now you should be able to access it. 

thank you for your support

**Comment by sevaa** [comment]
No immediate access. Requested.

**Comment by yashagarwal-314** [comment]
Hey Sevaa, 

unfortunately, I have no influence on the elf file, but this project is very important to me and It would be great if somehow you can guide me through a workaround, or you can make some changes in the library so the pyelftools works for this kind of files as well.

thank you!

**Comment by sevaa** [comment]
Is it possible that the file is intentionally obfuscated to prevent the kind of analysis you are trying to do?

**Comment by yashagarwal-314** [comment]
Hey Sevaa,

Thank you for your message! 

- **the elf file is the output of XC16 compiler** 
- Target is dspic33c core 
- its proprietary 16-bit architecture
- it is not obfuscated -> it can be read by xc16-readelf.exe 
- XC16 is based on the GCC v4.5.1 
- XC16 source code could be found here [https://ww1.microchip.com/downloads/en/DeviceDoc/v2.00.src.zip](url)

please let me know if you need any further information, thank you!

**Comment by sevaa** [comment]
If you have a version of readelf that can parse and dump the debug info, I suggest that you dump the dwarf info into a text file (use readelf -wi) and parse that. Should be enough for variable datatype recovery. You'll have to do some DIE ref chasing.

While it might be a fascinating project to figure out this flavor of DWARF, I don't think I can commit to that while not knowing the scope. Someone else might, but I don't see much enthusiasm here.

**Comment by eliben** [comment]
Closing this - parsing malformed DWARF that GNU tooling chokes on is not a task pyelftools is designed for

**Comment by sevaa** [comment]
Okay, some data points.

When the compiler vendors say theirs is a 16 bit machine, they take it seriously :) Looks like DWARF's standard [U]LEB128 integer encoding has been replaced with fixed width uint16 where possible.

My starting point was the abbrev table. Normally, one consists of mostly ULEB128 numbers - code, tag, (uint8) has-children flag followed by a set of (attribute, form) pairs until a null pair. Were the contents of the abbrev section be interpreted as all uint16's instead, the top of it looks like a sensibly looking abbrev:
```
Code 1
DW_TAG_compile_unit
Has children: yes
DW_AT_producer DW_FORM_string
DW_AT_language DW_FORM_data1 (but it's uint16 in the DIE anyway)
DW_AT_name DW_FORM_string
DW_AT_comp_dir DW_FORM_string
DW_AT_low_pc DW_FORM_addr
DW_AT_high_pc DW_FORM_addr
DW_AT_stmt_list DW_FORM_data4
0 0
```

Meanwhile in the info section, there is a UTF16 string starting with "GNU C..." at 0x18. Clearly the value of the producer attribute. Notably, it's preceded by uint16 0x1, which looks a lot like the abbrev code, followed by uint16 0x1 (DW_LANG_C89), then another UTF16 string that looks like a filename. Clearly DIE values.

That said, the format of their CU header is still something of a mystery to me. The standard CU header (before v5) is 0xb (11) bytes; this one is 0x16 (22), and it goes:

AC 00 03 00 00 00 00 00 02 00 00 00 00 00 00 00 00 00 00 00 04 00

Not sure what is what. AC 00 could be length, 03 00 could be DWARF version (but the doc says it emits DWARFv2), the next 4 bytes could be abbrev_offset.  Assuming they have promoted the uint8 address size to uint16 and that's what 02 00 is, there is a still bunch of numbers following it (5 uint16 zeros then a 4), with unclear meaning.

---

In conclusion, parsing this is definitely not a job for pyelftools :)

In theory, armed with this knowledge and with knowledge of DWARF proper, I could slap together an ad hoc parser that would dump the DIE tree for the OP. But I'm now wondering how far the OP is in their quest to recover the variable datatype from the readelf output, as I've suggested.

**Comment by sevaa** [comment]
@yashagarwal-314 see the latest on #518. Not sure about those odd nonzeros in your binary though.
